### PR TITLE
include: sys: util: implement `NUM_VA_ARGS`

### PR DESCRIPTION
**Pull request by ycsin**
This macro count the number of arguments in the variable arguments list.

Added test for it in `tests/lib/sys_util`.

[Doc preview](https://builds.zephyrproject.io/zephyr/pr/71890/docs/kernel/util/index.html#c.NUM_VA_ARGS_LESS_1)

### CODE REVIEW

**Review by pdgendt — DISMISSED** [review]
Nice

**Review by de-nordic — DISMISSED** [review]

File: include/zephyr/sys/util_internal.h [hunk 1]
```diff
@@ -136,6 +136,15 @@
 	_51, _52, _53, _54, _55, _56, _57, _58, _59, _60,	\
 	_61, _62, N, ...) N
 
+/* Implementation details for NUM_VA_ARGS */
+#define NUM_VA_ARGS_IMPL(...)                                                                      \
+	COND_CODE_1(                                                                               \
+		IS_EMPTY(__VA_ARGS__), (0),                                                        \
+		(Z_GET_ARG_64(__VA_ARGS__, 63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, \
```
Comment (de-nordic) [inline include/zephyr/sys/util_internal.h:143]: Question: Couldn't this part be replaced with `UTIL_INC(NUM_VA_ARGS_LESS_1(__VA_ARGS__))`?

Comment (ycsin) [reply include/zephyr/sys/util_internal.h:143]: Didn't try this myself, but the lowest value from `NUM_VA_ARGS_LESS_1` + `UTIL_INC` should be `1` whereas `NUM_VA_ARGS` is `0`

Comment (pdgendt) [reply include/zephyr/sys/util_internal.h:143]: The following appears to work:

```c
/* Implementation details for NUM_VA_ARGS */
#define NUM_VA_ARGS_IMPL(...)                                                                      \
	COND_CODE_1(                                                                               \
		IS_EMPTY(__VA_ARGS__), (0),                                                        \
		(UTIL_INC(NUM_VA_ARGS_LESS_1(__VA_ARGS__))))
```

**Review by pdgendt — DISMISSED** [review]

File: include/zephyr/sys/util_internal.h (hunk 1 above)
Comment (ycsin) [reply include/zephyr/sys/util_internal.h]: @pdgendt thanks for the suggestion, applied the change.
@de-nordic Sorry that I misunderstood your suggestion previously

**Review by peter-mitsis — DISMISSED** [review]

**Comment by ycsin** [comment]
- Did the implementation inplace in `include/zephyr/sys/util_macro.h` instead of `include/zephyr/sys/util_internal.h`, and
- Added test for `NUM_VA_ARGS_LESS_1`

@pdgendt @peter-mitsis please take another look, thanks.

**Review by pdgendt — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by andyross — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**